server/src/main/java/org/opensearch/ingest/IngestService.java

```diff
@@ -655,7 +655,7 @@ private void runBulkRequestInBatch(
     }
 
     private boolean shouldExecuteBulkRequestInBatch(int documentSize, int batchSize) {
-        return documentSize > 1 && batchSize > 1;
+        return batchSize > 1;
```

Comment:
There is branching logic in this class based off of this method, which in practice will return true in almost all cases assuming users don't specify a `batch_size` in their requests. Can we remove this branching and consolidate the logic?

Reply:
I removed this check, but there is still [branching logic](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/ingest/IngestService.java#L741) in `executePipelinesInBatchRequests` and when document size is 1, they are handled by `execute` function instead of `batchExecute`. Since even for single document index request `{index}/_doc/{docId}`, if an ingest pipeline is used, the above bulk logic is involved. I feel it's less risky to only change the call pattern of bulk API and keep the single doc indexing API untouched, WDYT?

Reply:
You're right that it is probably less risky, but I'm not sure that's a good reason. We need to have confidence that the new code path works, and it should work even in the case of a single document. This is a very similar situation to when the [asynchronous execute() method](https://github.com/opensearch-project/OpenSearch/blob/1219c568248fafa479d67a1eaa6e3e2d9748701e/server/src/main/java/org/opensearch/ingest/Processor.java#L69) was added: all code paths were changed to invoke the asynchronous method but a default implementation was included to delegate to existing the synchronous execute() method.

I'm okay not doing this refactoring in this PR if that's your preference, but I think we should still do that simplification. We might need to do more testing to give us confidence the existing behavior doesn't actually change to mitigate the risk you mentioned. What do you think?

Reply:
Agreed. Let's have this PR to only change default behavior of _bulk API, and change single doc ingest code path in 2.17 and 3.x. I'll add more test coverage in the following PRs.

Reply:
Let me know if you have any other comments

Reply:
@andrross could you take a look at this PR again once available?